builtin/providers/aws/resource_aws_security_group.go

```diff
@@ -876,3 +878,68 @@ func protocolForValue(v string) string {
 	log.Printf("[WARN] Unable to determine valid protocol: no matching protocols found")
 	return protocol
 }
+
+func getLambdaENIs(conn *ec2.EC2, d *schema.ResourceData) []ec2.NetworkInterface {
+	filter := &ec2.Filter{
+		Name:   aws.String("group-id"),
+		Values: []*string{aws.String(d.Id())},
+	}
+	params := &ec2.DescribeNetworkInterfacesInput{
+		Filters: []*ec2.Filter{filter},
+	}
+	networkInterfaceResp, err := conn.DescribeNetworkInterfaces(params)
+
+	lambdaENIs := make([]ec2.NetworkInterface, 0)
+	if err != nil {
+		log.Printf("[WARN] Unable to fetch ENIs: %s", err)
+		return lambdaENIs
+	}
+
+	v := networkInterfaceResp.NetworkInterfaces
+	for _, eni := range v {
+		if strings.Contains(*eni.Description, "AWS Lambda VPC ENI") {
+			lambdaENIs = append(lambdaENIs, *eni)
+		}
+	}
+	return lambdaENIs
+}
+
+func deleteLingeringLambdaENIs(conn *ec2.EC2, d *schema.ResourceData) error {
+	lambdaENIs := getLambdaENIs(conn, d)
+
+	for i := 0; i < len(lambdaENIs); i++ {
```

Comment:
I think this and the next for loop can be merged into a single one and shouldn't even become too messy if you leverage the existing retry helpers `resource.StateChangeConf` like [here](https://github.com/hashicorp/terraform/blob/master/builtin/providers/aws/resource_aws_network_interface.go#L204-L214).

That will also allow us to error out if we _actually_ failed to detach the network interface for any weird reason (i.e. if the ENI fails to detach after given `timeout`). Such edge case is currently ignored in this implementation as you just _hope_ it will be detached after 1 second.

Potentially you may be able to squeeze the third loop w/ `DeleteNetworkInterface` here too as that would also reduce complexity (from `3n` to `n`)... but I don't see that as a big deal compared to the first two loops.

Either way in all of these loops there is no need to use the C-like approach with index-incrementation, instead you can easily [range](https://github.com/golang/go/wiki/Range) over the slice:

```go
for _, eni := range lambdaENIs {
  params := &ec2.DetachNetworkInterfaceInput{
    AttachmentId: eni.Attachment.AttachmentId,
  }
```